abi-versions/vNEXT/README.md

```diff
@@ -0,0 +1,919 @@
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Cancel outstanding gRPC request or stream (`callout_id`).
+
+
+### `proxy_close_grpc_call`
+
+* params:
+  - `i32 (uint32_t) callout_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Cancel outstanding gRPC request or close existing gRPC stream (`callout_id`).
+
+
+## Shared Key-Value store
+
+### `proxy_get_shared_data`
+
+* params:
+  - `i32 (const char*) key_data`
+  - `i32 (size_t) key_size`
+  - `i32 (const char**) return_value_data`
+  - `i32 (size_t*) return_value_size`
+  - `i32 (uint32_t*) return_cas`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Get shared data identified by a key (`key_data`, `key_size`). The compare-and-switch value
+(`return_cas`) is returned and can be used when updating the value with `proxy_set_shared_data`.
+
+
+### `proxy_set_shared_data`
+
+* params:
+  - `i32 (const char*) key_data`
+  - `i32 (size_t) key_size`
+  - `i32 (const char*) value_data`
+  - `i32 (size_t) value_size`
+  - `i32 (uint32_t) cas`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Set shared data identified by a key (`key_data`, `key_size`) to a value (`value_data`,
+`value_size`). If compare-and-switch value (`cas`) is set, then it must match the current value in
+order to update to succeed.
+
+
+## Shared queue
+
+### `proxy_register_shared_queue`
+
+* params:
+  - `i32 (const char*) queue_name_data`
+  - `i32 (size_t) queue_name_size`
+  - `i32 (uint32_t*) return_queue_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Register a shared queue using a given name (`queue_name_data`, `queue_name_size`). It can be
+referred to in `proxy_enqueue_shared_queue` and `proxy_dequeue_shared_queue` using returned unique
+queue identifier (`queue_id`).
+
+
+### `proxy_resolve_shared_queue`
+
+* params:
+  - `i32 (const char*) queue_name_data`
+  - `i32 (size_t) queue_name_size`
+  - `i32 (uint32_t*) return_queue_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Resolves existing shared queue using a given name (`queue_name_data`, `queue_name_size`). It can be
+referred to in `proxy_enqueue_shared_queue` and `proxy_dequeue_shared_queue` using returned unique
+queue identifier (`queue_id`).
+
+
+### `proxy_dequeue_shared_queue`
+
+* params:
+  - `i32 (uint32_t) queue_id`
+  - `i32 (const char**) payload_data`
+  - `i32 (size_t*) payload_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Get data (`payload_data`, `payload_size`) to the end of the queue (`queue_id`).
+
+
+### `proxy_enqueue_shared_queue`
+
+* params:
+  - `i32 (uint32_t) queue_id`
+  - `i32 (const char*) payload_data`
+  - `i32 (size_t) payload_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Add data (`payload_data`, `payload_size`) to the front of the queue (`queue_id`).
+
+
+### `proxy_remove_shared_queue`
+
+* params:
+  - `i32 (uint32_t) queue_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Remove queue (`queue_id`).
+
+
+## Stats/metrics
```

Comment:
I've commented on this elsewhere but I *very* strongly recommend that you split counters/gauges/histograms into independent APIs. The reason is that they are all subtly different. If we don't want to do this, I think there are a few things we need to do here:
1) Better clarify what call patterns will result in invalid arguments (incrementing a counter with a negative offset, trying to set a counter value, etc.).
2) Figure out how to deal with histograms correctly because `proxy_get_metric` is not going to work well for them.

Reply:
It isn't hard, it is just much less flexible.  If we add another metric type then we have to change the ABI.  Also, the Istio team currently uses another layer which abstracts the different metrics so that they can use a configuration file to define them.  The fact that they can just call proxy_record_metric greatly simplifies the implementation and was a big boon for them.  Given that this is a low level ABI, there is no value IMO in splitting them and great value in keeping it general like it is now.

Reply:
I think we should cover this in our video meeting tomorrow, but briefly, I think it's a false simplification to try to jam all of the different metric types into a single API. It might be possible for counters/gauges (though even that is a bit strange) but it goes totally off the rails for histograms or histogram like things (e.g. the ATS thing you mentioned). I'm not sure how sensible it is to try to get histograms/other into this API.

Reply:
I disagree.  For histograms you can get a single Y value by providing one additional argument which is the X value.  I don't think we need serialization.  The bucketing should be known through other channels.  It can be in the config for example.